### ISSUE TITLE
removeClass()

### ISSUE THREAD

**Issue by rosell-dk** [HubSpot/youmightnotneedjquery#170]
The current implementation of .removeClass() is not optimal.

A major problem is that it may replace a part of a class name in class names that contains a dash. The problem lies in the \b, which treats dashes as word boundaries.

Another problem is that it does not behave consistently across browsers. This is because the regex doesn't behave quite the same way as el.classList.remove(). For example, el.classList.remove does not ignore case, while the regex does. Also, el.classList.remove doesn't end up with extra spaces, while the regex does. Also, strangely, the regex is created to support removal of multiple class names (only that part is buggy), while el.classList.remove does not support removal of multiple class names in one go.

Such inconsistencies can result in bugs that requires cross-browser testing to discover.
For that reason, I suggest that the el.classList.remove() part is removed. Instead, I propose it is added as an IE10+ alternative.

Also, the current regex can be optimized, because "\b" actually also matches beginning (^) and end ($).

The following implementation addresses all these issues:

**Remove single classname, IE8+**

```
function removeClass(el, className) {
  el.className=el.className.replace(new RegExp("(\\s|^)"+className+"(?!\\S)","g"),"").trim()
}
```

For IE9+, there is a more elegant solution, which also overcomes all those issues:
**Remove single classname, IE9+**

```
function removeClass​(el, className​) {
  el​.className=el​.className​.split​(/\s/​)​.filter​(function​(c​){return c!=className;}​)​.join​(" "​);
}
```

Another thing is that with jQuery, you can remove multiple class names in one go.
I'm not sure you want that. It seems a little bloat-ish to me. Also, el.classList.remove() doesn't allow that.

But if you do want to support removal of multiple class names in one go, here are some implementations, which also removes extra spaces:

**Remove multiple classnames, IE8+**:

```
function removeClass​(el, className​) {
  el​.className = el​.className​.replace​(new RegExp​("​(^|\\s+​)" + className​.replace​(/\s/g,"​(?!\\S​)|​(^|\\s+​)"​) + "​(?!\\S​)","g"​),""​)​.trim​(​);
}
```

**Remove multiple classnames, IE9+**:

```
function removeClass​(el, className​) {
  el.className = el.className.split(/\s/).filter(function(c){return 0>className.split(" ").indexOf(c);}).join(" ");
}
```

I have created compliance tests, which are available here: http://picoquery.com/lab/youmightnotneedjquery-tests/removeclass/

By the way, regarding the extra spaces, it is no problem with in the sense, that the browsers does not behave strangely. The potential problem arises, if the programmer later compares the classname directly. For example, this code may behave different than intended: `removeClass(el, "cancelled"); if (el.className == "") {...}`. If you should decide that extra spaces do not need to be removed, the IE8+ versions can be a bit shorter. I have these variants available in the compliance tests
http://picoquery.com/lab/youmightnotneedjquery-tests/removeclass/
http://picoquery.com/lab/youmightnotneedjquery-tests/removeclass-multiple/

**Comment by imme-emosol** [comment]
Could this be a _Remove multiple classnames, No unwanted spaces-version, IE8+_ ?
`element.className = element.className.replace( new RegExp( "\\b" + className.replace( " " , "\\b|\\b" ) + "\\b" , "g" ) , " " ).replace( /\s\s/ , " " ).trim();`

**Comment by rosell-dk** [comment]
Actually, when multiple classnames are removed, there is the case that if two classnames next to each other are removed, the string will contain 3 spaces. So in this case of removing multiple classnames, we need to reduce 2+ spaces to single space. Instead of matching /\s\s/ we can match this pattern: /\s+/. And also, there can be multiple wholes, so we must use the global flag: /\s+/g

This should work:
`element.className = element.className.replace( new RegExp( "\\b" + className.replace( " " , "\\b|\\b" ) + "\\b" , "g" ) , " " ).replace( /\s+/g, " " ).trim();`

I haven't tested it yet, though.
I will set up some unit-tests in the beginning of next week and post a link to them here, so we can all be reassured that the code works

**Comment by rosell-dk** [comment]
I have updated the original comment in order not to have suboptimal solutions listed. Also, I now have completed the compliance-tests, which are available here:  http://picoquery.com/lab/youmightnotneedjquery-tests/removeclass/

**Comment by Hum4n01d** [comment]
I'm not a contributor to this project, but I'd suggest that you make a PR with your proposed changes

**Comment by Hum4n01d** [comment]
Well what do you know! #109

**Comment by Abbondanzo** [comment]
Closed by #243
